File: Rust/Lexicographically_Smallest_String_After_Applying_Operations.rs
use std::collections::{HashSet , VecDeque};
impl Solution {
    pub fn find_lex_smallest_string(s: String, a: i32, b: i32) -> String {
        let mut done = HashSet::new();
        let mut q = VecDeque::new();
        done.insert(s.clone());
        q.push_back(s.clone());
        let mut ans = s;
        while let Some(curr) = q.pop_front(){
            if curr < ans{
                ans = curr.clone();
            }
            let mut chars: Vec<char> = curr.chars().collect();
            for i in (1..chars.len()).step_by(2){
            let digit = chars[i].to_digit(10).unwrap() as i32;
            let new_d = ((digit + a ) % 10) as u32;
            chars[i] = std::char::from_digit(new_d, 10).unwrap();
            }
            let added:String = chars.iter().collect();
            if done.insert(added.clone()){
                q.push_back(added);
            }
            let n = curr.len();
            let b = b as usize;
            let rotated = format!{
                "{}{}",
                &curr[n - b..],
                &curr[..n - b]
            };
            if done.insert(rotated.clone()){
                q.push_back(rotated);
            }
        }
        ans
    }
}

File: Cpp/Lexicographically_Smallest_String_After_Applying_Operations.cpp
class Solution {
public:
    string findLexSmallestString(string s, int a, int b) {
        unordered_set<string> done;
        queue<string> q;
        done.insert(s);
        q.push(s);
        string ans = s;
        while (!q.empty()) {
            string curr = q.front();
            q.pop();
            if (curr < ans) {
                ans = curr;
            }
            string added = curr;
            for (int i = 1; i < added.size(); i += 2) {
                int digit = added[i] - '0';
                digit = (digit + a) % 10;
                added[i] = digit + '0';
            }
            if (!done.count(added)) {
                done.insert(added);
                q.push(added);
            }
            int n = curr.size();
            string rotated = curr.substr(n - b) + curr.substr(0, n - b);
            if (!done.count(rotated)) {
                done.insert(rotated);
                q.push(rotated);
            }
        }
        return ans;
    }
};